onnxruntime/contrib_ops/cpu/quantization/qlinear_softmax.cc

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "qlinear_softmax.h"
+#include <cstddef>
+#include <cstdint>
+#include "core/common/common.h"
+#include "core/providers/common.h"
+#include "core/providers/cpu/tensor/transpose.h"
+
+#include "core/mlas/inc/mlas.h"
+#include "core/util/math.h"
+#include "core/util/math_cpuonly.h"
+#include "core/platform/threadpool.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+void QlinearBuildLookupTableUint32(uint32_t* table,
+                                   const float x_scale,
+                                   size_t reduce_len) {
+  const double qscale =
+      fmin(((double)UINT32_MAX) / (double)reduce_len, double(0x7fffff));
+  for (int32_t i = 0; i < 256; i++) {
+    const double scaled_exp_xi =
+        qscale * exp((double)(i - 255) * (double)x_scale);
+    table[(uint32_t)i] = (uint32_t)lrint(scaled_exp_xi);
+  }
+}
+
+template <typename T>
+void QLinearSoftmax<T>::BuildLookupTableIfFixed(const OpKernelInfo& info, uint32_t channels) {
+  const Tensor* tensor_x_scale = nullptr;
+  const Tensor* tensor_x_zero_point = nullptr;
+  const Tensor* tensor_y_scale = nullptr;
+  const Tensor* tensor_y_zero_point = nullptr;
+
+  bool get_x_scale = info.TryGetConstantInput(1, &tensor_x_scale);
+  bool get_x_zero_point = !info.node().InputDefs()[2]->Exists() || info.TryGetConstantInput(2, &tensor_x_zero_point);
+  bool get_y_scale = info.TryGetConstantInput(3, &tensor_y_scale);
+  bool get_y_zero_point = !info.node().InputDefs()[4]->Exists() || info.TryGetConstantInput(4, &tensor_y_zero_point);
+  bool is_fixed_parameters = get_x_scale && get_x_zero_point && get_y_scale && get_y_zero_point;
+  ORT_ENFORCE(IsScalarOr1ElementVector(tensor_x_scale),
+              "QlinearBuildLookupTable : input X_scale must be a scalar or 1D tensor of size 1");
+
+  if (is_fixed_parameters) {
+    fixed_lookup_table_.resize(256);
+    const float X_scale = *(tensor_x_scale->Data<float>());
+    QlinearBuildLookupTableUint32(fixed_lookup_table_.data(), X_scale, channels);
+  }
+}
+
+template <typename T>
+QLinearSoftmax<T>::QLinearSoftmax(const OpKernelInfo& info)
+    : OpKernel(info) {
+  const auto& node = info.node();
+  opset_ = node.SinceVersion();
```

Comment:
>pset_ = node.SinceVersion(); [](http://example.com/codeflow?start=3&length=28)

SinceVersion will be 1. This is different with Softmax.

Reply:
we can handle the opset in the quantization tool or QDQ fusion part by always adding axis.

Reply:
Do you mean considering OPset13 or above is enough in the implementation? Since Opset12 or below has different meaning with axis.

Reply:
Add a new attribute "opset" in the schema of QlinearSoftmax